oceanspy/utils.py

```diff
@@ -586,3 +606,439 @@ def Coriolis_parameter(Y, omega=7.2921e-5):
     e = 2 * omega * _np.cos(Y_rad)
 
     return f, e
+
+
+@njit
+def spherical2cartesian_compiled(Y, X, R=6371.0):
+    """
+    Convert spherical coordinates to cartesian.
+    Parameters
+    ----------
+    Y: np.array
+        Spherical Y coordinate (latitude)
+    X: np.array
+        Spherical X coordinate (longitude)
+    R: scalar
+        Earth radius in km
+        If None, use geopy default
+    Returns
+    -------
+    x: np.array
+        Cartesian x coordinate
+    y: np.array
+        Cartesian y coordinate
+    z: np.array
+        Cartesian z coordinate
+    """
+
+    # Convert
+    Y_rad = np.deg2rad(Y)
+    X_rad = np.deg2rad(X)
+    x = R * np.cos(Y_rad) * np.cos(X_rad)
+    y = R * np.cos(Y_rad) * np.sin(X_rad)
+    z = R * np.sin(Y_rad)
+
+    return x, y, z
+
+
+@njit
+def to_180(x):
+    """
+    convert any longitude scale to [-180,180)
+    """
+    x = x % 360
+    return x + (-1) * (x // 180) * 360
+
+
+def local_to_latlon(u, v, cs, sn):
+    """convert local vector to north-east"""
+    uu = u * cs - v * sn
+    vv = u * sn + v * cs
+    return uu, vv
+
+
+def get_combination(lst, select):
+    """
+    Iteratively find all the combination that
+    has (select) amount of elements
+    and every element belongs to lst
+    """
+    if select == 1:
+        return [[num] for num in lst]
+    else:
+        the_lst = []
+        for i, num in enumerate(lst):
+            sub_lst = get_combination(lst[i + 1 :], select - 1)
+            for com in sub_lst:
+                com.append(num)
+            #             print(sub_lst)
+            the_lst += sub_lst
+        return the_lst
+
+
+def grid2array(od, all_of_them=False):
+    """
+    read the od grid things into numpy arrays
+    stored in this module
+    """
+    global Z, Zl, dZ, dZl, dXC, dYC, dXG, dYG, XC, XG, YC, YG, CS, SN, tree, ts
```

Comment:
We should try to avoid global variables as much as possible (or it would be better to use python's cache decorators)